src/coreclr/vm/multicorejit.cpp

```diff
@@ -488,75 +491,59 @@ HRESULT MulticoreJitRecorder::WriteOutput(IStream * pStream)
+                DWORD data1 = m_JitInfoArray[i].data1;
+                _ASSERTE(data1 != 0);
+                hr = WriteData(pStream, &data1, sizeof(data1));
             }
-
-            dwData = Pack8_24(MULTICOREJIT_JITINF_RECORD_ID,  count * sizeof(DWORD) + sizeof(DWORD));
-
-            hr = WriteData(pStream, & dwData, sizeof(dwData));
-
-            if (SUCCEEDED(hr))
+            else
             {
-                hr = WriteData(pStream, pInfo, sizeof(unsigned) * count);
-            }
-
-            pInfo  += count;
-            remain -= count;
-        }
-    }
+                // Method record
+                DWORD data1 = m_JitInfoArray[i].data1;
+                DWORD data2 = m_JitInfoArray[i].data2;
+                BYTE * pSignature = (BYTE *) m_JitInfoArray[i].ptr;
 
-    if (SUCCEEDED(hr) && genericInfoArray != nullptr)
-    {
-        for (LONG i = 0 ; i < m_GenericInfoCount && SUCCEEDED(hr); i++)
-        {
-            unsigned info = genericInfoArray[i].genericInfo;
-            BYTE * pSignature = genericInfoArray[i].genericSignature;
-            if (info == 0 && pSignature == nullptr)
-            {
-                continue;
-            }
+                if (pSignature == nullptr)
+                {
+                    // Skipped method
+                    continue;
+                }
 
-            DWORD sigSize = info & 0xFFFFFF;
-            DWORD dataSize = sigSize * sizeof(BYTE) + sizeof(DWORD) * 2;
-            DWORD dwSize = ((DWORD)(dataSize + sizeof(DWORD) - 1) / sizeof(DWORD)) * sizeof(DWORD);
-            _ASSERTE(dwSize <= 0xFFFFFF);
-            dwData = Pack8_24(MULTICOREJIT_GENERICINF_RECORD_ID, dwSize);
-            hr = WriteData(pStream, &dwData, sizeof(dwData));
-            if (SUCCEEDED(hr))
-            {
-                hr = WriteData(pStream, &info, sizeof(unsigned));
-            }
+                _ASSERTE(data1 != 0);
+                _ASSERTE(data2 != 0);
+                _ASSERTE(pSignature != nullptr);
 
-            if (SUCCEEDED(hr))
-            {
-                hr = WriteData(pStream, pSignature, sizeof(BYTE) * sigSize);
-            }
+                DWORD sigSize = data2 & (MAX_SIGNATURE_LENGTH - 1);
+                DWORD dataSize = sigSize + sizeof(DWORD) * 2;
+                DWORD dwSize = ((DWORD)(dataSize + sizeof(DWORD) - 1) / sizeof(DWORD)) * sizeof(DWORD);
+                _ASSERTE(dwSize < MAX_SIGNATURE_LENGTH);
+                data2 = dwSize << SIGNATURE_LENGTH_OFFSET | data2;
 
-            if (SUCCEEDED(hr))
-            {
-                DWORD init = 0;
-                hr = WriteData(pStream, &init, dwSize - dataSize);
+                hr = WriteData(pStream, &data1, sizeof(data1));
+                if (SUCCEEDED(hr))
+                {
+                    hr = WriteData(pStream, &data2, sizeof(data2));
+                }
+                if (SUCCEEDED(hr))
+                {
+                    hr = WriteData(pStream, pSignature, sigSize);
```

Comment:
I wanted to unify approach here and not have two distinct ways of dealing with methods. I agree that it's better to split generics/non-generics, because profile size increases too much according to my measurements.